### ISSUE TITLE
Incorrect suggestion made by question_mark when match ergonomics involved

### ISSUE THREAD

**Issue by Plecra** [rust-lang/rust-clippy#11983]
### Summary

Possibly related to https://github.com/rust-lang/rust-clippy/issues/11764

### Reproducer

I tried this code:

```rs
fn foo() -> Option<()> {
let alias = Some(String::new());
let Some(v) = &alias else { return None };
let _ = v;
Some(())
}
```

I expected to see this happen:

A valid suggestion would be produced. This is non-trivial, since clippy wants to remove the match. `as_ref()` works for `Option`s, but not in the general case

Instead, this happened:

```rs
warning: this `let...else` may be rewritten with the `?` operator
 --> src/main.rs:4:1
  |
4 | let Some(v) = &alias else { return None };
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `let v = &alias?;`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
  = note: `#[warn(clippy::question_mark)]` on by default

warning: `playground` (bin "playground") generated 1 warning
```

**note:** The suggestion doesn't compile as it has different semantics, and applies the `Try` operator before taking the reference.

### Version

On the playground, which indicates version 0.1.74 (2023-12-04 a28077b)

**Comment by ronnodas** [comment]
The suggestion [seems to compile fine](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8f78bd520b8b758f9bbd5eb4129c7791)? Why do you say it has different semantics?

**Comment by cvijdea-bd** [comment]
Might be related to https://github.com/rust-lang/rust-clippy/issues/12412 (or even a duplicate)

**Comment by J-ZhengLi** [comment]
Seems it was added from this [PR](https://github.com/rust-lang/rust-clippy/pull/11994)

**Comment by J-ZhengLi** [comment]
Reproducer that actually fail to compile:

```rust
fn foo(alias: &Option<String>) -> Option<()> {
    let Some(v) = &alias else { return None };
    // let v = &alias?;
    Some(())
}
```
